File: bindings/lua/commands.lua
local uhppote = require("src/uhppote")
local os = require("os")

local CONTROLLER <const> = 405419896
local ADDRESS <const> = "192.168.1.100"
local NETMASK <const> = "255.255.255.0"
local GATEWAY <const> = "192.168.1.1"
local LISTENER <const> = { ["address"] = "192.168.1.100", ["port"] = 60001 }
local DOOR = 1
local DOOR_MODE = "controlled"
local DOOR_DELAY = 5
local CARD = 10058400
local CARD_INDEX = 1
local CARD_START_DATE = "2023-01-01"
local CARD_END_DATE = "2023-12-31"
local CARD_DOORS = "1,2,3,4"
local CARD_PIN = 0
local EVENT_INDEX = 37
local TIME_PROFILE_ID = 29

local TASKS = {
    [0] = "door-controlled",
    [1] = "door-normally-open",
    [2] = "door-normally-closed",
    [3] = "disable-time-profile",
    [4] = "enable-time-profile",
    [5] = "card-no-password",
    [6] = "card-in-password",
    [7] = "card+password",
    [8] = "enable-more-cards",
    [9] = "disable-more-cards",
    [10] = "trigger-once",
    [11] = "disable-pushbutton",
    [12] = "enable-pushbutton"
}

local INTERLOCKS = {
    [0] = "none",
    [1] = "1&2",
    [2] = "3&4",
    [3] = "1&2,3&4",
    [4] = "1&2&3",
    [8] = "1&2&3&4",
}

function get_all_controllers(args)
    return uhppote.get_all_controllers()
end

function get_controller(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.get_controller(controller)
end

function set_ip(args)
    local controller = parse(args,"controller",CONTROLLER)
    local address = parse(args,"address",ADDRESS)
    local netmask = parse(args,"netmask",NETMASK)
    local gateway = parse(args,"gateway",GATEWAY)

    return uhppote.set_ip(controller,address,netmask,gateway)
end

function get_time(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.get_time(controller)
end

function set_time(args)
    local controller = parse(args,"controller",CONTROLLER)
    local datetime = parse(args,"time",os.date("%Y-%m-%d %H:%M:%S"))

    return uhppote.set_time(controller,datetime)
end

function get_status(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.get_status(controller)
end

function get_listener(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.get_listener(controller)
end

function set_listener(args)
    local controller = parse(args,"controller",CONTROLLER)
    local address = parse(args,"address",LISTENER.address)
    local port = parse(args,"port",LISTENER.port)

    return uhppote.set_listener(controller,address,port)
end

function get_door_control(args)
    local controller = parse(args, "controller", CONTROLLER)
    local door = parse(args, "door", DOOR)

    return uhppote.get_door_control(controller, door)
end

function set_door_control(args)
    local controller = parse(args, "controller", CONTROLLER)
    local door = parse(args, "door", DOOR)
    local mode = parse(args, "mode", DOOR_MODE)
    local delay = parse(args, "delay", DOOR_DELAY)

    if mode == "normally-open" then
        mode = 1
    elseif mode == "normally-closed" then
        mode = 2
    elseif mode == "controlled" then
        mode = 3
    end

    return uhppote.set_door_control(controller, door, mode, delay)
end

function open_door(args)
    local controller = parse(args, "controller", CONTROLLER)
    local door = parse(args,"door",DOOR)

    return uhppote.open_door(controller, door)
end

function get_cards(args) 
    local controller = parse(args,"controller", CONTROLLER)

    return uhppote.get_cards(controller)
end

function get_card(args)
    local controller = parse(args, "controller", CONTROLLER)
    local card = parse(args, "card", CARD)

    return uhppote.get_card(controller, card)
end

function get_card_by_index(args)
    local controller = parse(args, "controller", CONTROLLER)
    local index = parse(args, "index", CARD_INDEX)

    return uhppote.get_card_by_index(controller, index)
end

function put_card(args)
    local controller = parse(args, "controller", CONTROLLER)
    local card = parse(args, "card", CARD)
    local start_date = parse(args, "start_date", CARD_START_DATE)
    local end_date = parse(args, "end_date", CARD_END_DATE)
    local doors = parse(args, "doors", CARD_DOORS)
    local PIN = parse(args, "PIN", CARD_PIN)

    local permissions = { 0,0,0,0 }
    for v in doors:gmatch("([^, ]+)") do
        local door, profile = v:match("^([1234]):(%d+)$")

        if door and profile then
          permissions[tonumber(door)] = tonumber(profile)
        elseif v:match("^[1234]$") then
          permissions[tonumber(v)] = 1
        end
    end

    return uhppote.put_card(controller, 
                            card, 
                            start_date, end_date, 
                            permissions[1], permissions[2], permissions[3], permissions[4], 
                            PIN)
end

function delete_card(args)
    local controller = parse(args,"controller",CONTROLLER)
    local card = parse(args,"card",CARD)

    return uhppote.delete_card(controller, card)
end

function delete_all_cards(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.delete_all_cards(controller)
end

function get_event(args)
    local controller = parse(args,"controller",CONTROLLER)
    local index = parse(args,"index",EVENT_INDEX)

    return uhppote.get_event(controller, index)
end

function get_event_index(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.get_event_index(controller)
end

function set_event_index(args)
    local controller = parse(args,"controller",CONTROLLER)
    local index = parse(args,"index",EVENT_INDEX)

    return uhppote.set_event_index(controller, index)
end

function record_special_events(args)
    local controller = parse(args,"controller",CONTROLLER)
    local disabled = parse(args,"disabled",false)

    return uhppote.record_special_events(controller, not disabled)
end

function get_time_profile(args)
    local controller = parse(args,"controller",CONTROLLER)
    local profile = parse(args,"profile",TIME_PROFILE_ID)

    return uhppote.get_time_profile(controller, profile)
end

function set_time_profile(args)
    local controller = parse(args,"controller",CONTROLLER)
    local profile = parse(args,"profile",TIME_PROFILE_ID)
    local start_date = parse(args,"start_date","2023-01-01")
    local end_date = parse(args,"end_date","2099-01-01")
    local linked_profile = tonumber(parse(args,"linked","0"))

    local weekdays = string.lower(parse(args, "weekdays", ""))
    local monday = string.match(weekdays,"mon")
    local tuesday = string.match(weekdays,"tue")
    local wednesday = string.match(weekdays,"wed")
    local thursday = string.match(weekdays,"thu")
    local friday = string.match(weekdays,"fri")
    local saturday = string.match(weekdays,"sat")
    local sunday = string.match(weekdays,"sun")

    local segments = { { from = "00:00", to = "00:00" },
                       { from = "00:00", to = "00:00" },
                       { from = "00:00", to = "00:00" } }

    local ix = 1
    for p,q in string.gmatch(parse(args, "segments", ""), "(%d?%d:%d%d)%-(%d?%d:%d%d)") do
        segments[ix].from = p
        segments[ix].to = q
        ix = ix + 1
     end

    return uhppote.set_time_profile(controller, profile,
                                    start_date, end_date,
                                    monday, tuesday, wednesday, thursday, friday, saturday, sunday,
                                    segments[1].from, segments[1].to,
                                    segments[2].from, segments[2].to,
                                    segments[3].from, segments[3].to,
                                    linked_profile)
end

function delete_all_time_profiles(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.delete_all_time_profiles(controller)
end

function add_task(args)
    local controller = parse(args,"controller",CONTROLLER)
    local task = parse(args,"task",2)
    local start_date = parse(args,"start_date","2023-01-01")
    local end_date = parse(args,"end_date","2099-01-01")
    local at = string.match(parse(args, "at", "00:00"),"(%d?%d:%d%d)") or "12:34"
    local door = tonumber(parse(args,"door","0"))
    local weekdays = string.lower(parse(args, "weekdays", ""))
    local more_cards = tonumber(parse(args,"more_cards","0"))

    local monday = string.match(weekdays,"mon")
    local tuesday = string.match(weekdays,"tue")
    local wednesday = string.match(weekdays,"wed")
    local thursday = string.match(weekdays,"thu")
    local friday = string.match(weekdays,"fri")
    local saturday = string.match(weekdays,"sat")
    local sunday = string.match(weekdays,"sun")
    local task = string.lower(parse(args, "task", ""))

    for k,v in pairs(TASKS) do
      if v == task then
          return uhppote.add_task(controller, 
                                  start_date, end_date,
                                  monday, tuesday, wednesday, thursday, friday, saturday, sunday,
                                  at,
                                  door,
                                  k,
                                  more_cards)
      end
    end

    error("invalid task")
end

function refresh_tasklist(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.refresh_tasklist(controller)
end

function clear_tasklist(args)
    local controller = parse(args,"controller",CONTROLLER)

    return uhppote.clear_tasklist(controller)
end

function set_pc_control(args)
    local controller = parse(args,"controller",CONTROLLER)
    local disabled = parse(args,"disabled",false)

    return uhppote.set_pc_control(controller, not disabled)
end

function set_interlock(args)
    local controller = parse(args,"controller",CONTROLLER)
    local interlock = parse(args, "interlock", "")

    for k,v in pairs(INTERLOCKS) do
      if v == interlock then
          return uhppote.set_interlock(controller, k)
      end
    end

    error ("invalid interlock")
end

function activate_keypads(args)
    local controller = parse(args,"controller",CONTROLLER)
    local keypads = parse(args, "keypads", "")

    local reader1 = false
    local reader2 = false
    local reader3 = false
    local reader4 = false

    for reader in string.gmatch(keypads, "([1234])[, ]*") do
        if reader == "1" then
          reader1 = true
        elseif reader == "2" then
          reader2 = true
        elseif reader == "3" then
          reader3 = true
        elseif reader == "4" then
          reader4 = true
        end
    end 

    return uhppote.activate_keypads(controller, reader1, reader2, reader3, reader4)
end

function set_door_passcodes(args)
    local controller = parse(args,"controller",CONTROLLER)
    local door = parse(args, "door", DOOR)
    local passcodes = parse(args, "passcodes", "")

    local codes = { 0, 0, 0, 0 }
    local ix = 1
    for v in string.gmatch(passcodes, "([0-9]+)[, ]*") do
      local code = tonumber(v)

      if code > 0 and code < 1000000 then
         codes[ix] = code
         ix = ix + 1
      end
    end 

    return uhppote.set_door_passcodes(controller, tonumber(door), codes[1], codes[2], codes[3], codes[4])
end

function listen(args)
    local onerror = function(err)
                       print("   *** ERROR", err)
                   end

    local handler = function(event)
                       pprint(event)
                   end

    -- Ref. https://stackoverflow.com/questions/50506099/why-is-lua-not-respecting-my-sigint-signal
    -- posix.signal(posix.SIGINT, function(signum)
    --     os.exit(0)
    -- end)
    -- 
    -- posix.signal(posix.SIGTERM, function(signum)
    --     os.exit(0)
    -- end)

    return uhppote.listen(handler, onerror)
end

-- stylua: ignore start
local commands = {
   commands = {     
       { ["command"] = "get-all-controllers",      ["f"] = get_all_controllers,      flags = {},             options = {} },
       { ["command"] = "get-controller",           ["f"] = get_controller,           flags = {},             options = { "controller" } },
       { ["command"] = "set-ip",                   ["f"] = set_ip,                   flags = {},             options = { "controller","address","netmask","gateway" } },
       { ["command"] = "get-time",                 ["f"] = get_time,                 flags = {},             options = { "controller" } },
       { ["command"] = "set-time",                 ["f"] = set_time,                 flags = {},             options = { "controller","time" } },
       { ["command"] = "get-listener",             ["f"] = get_listener,             flags = {},             options = { "controller" } },
       { ["command"] = "set-listener",             ["f"] = set_listener,             flags = {},             options = { "controller","address","port" } },
       { ["command"] = "get-status",               ["f"] = get_status,               flags = {},             options = { "controller" } },
       { ["command"] = "get-door-control",         ["f"] = get_door_control,         flags = {},             options = { "controller","door" } },
       { ["command"] = "set-door-control",         ["f"] = set_door_control,         flags = {},             options = { "controller","door","mode","delay" } },
       { ["command"] = "open-door",                ["f"] = open_door,                flags = {},             options = { "controller","door" } },
       { ["command"] = "get-cards",                ["f"] = get_cards,                flags = {},             options = { "controller" } },
       { ["command"] = "get-card",                 ["f"] = get_card,                 flags = {},             options = { "controller","card" } },
       { ["command"] = "get-card-by-index",        ["f"] = get_card_by_index,        flags = {},             options = { "controller","index" } },
       { ["command"] = "put-card",                 ["f"] = put_card,                 flags = {},             options = { "controller","card","start-date","end-date","doors","PIN" } },
       { ["command"] = "delete-card",              ["f"] = delete_card,              flags = {},             options = { "controller","card" } },
       { ["command"] = "delete-all-cards",         ["f"] = delete_all_cards,         flags = {},             options = { "controller" } },
       { ["command"] = "get-event",                ["f"] = get_event,                flags = {},             options = { "controller","index" } },
       { ["command"] = "get-event-index",          ["f"] = get_event_index,          flags = {},             options = { "controller" } },
       { ["command"] = "set-event-index",          ["f"] = set_event_index,          flags = {},             options = { "controller", "index" } },
       { ["command"] = "record-special-events",    ["f"] = record_special_events,    flags = { "disabled" }, options = { "controller" } },
       { ["command"] = "get-time-profile",         ["f"] = get_time_profile,         flags = {},             options = { "controller", "profile" } },
       { ["command"] = "set-time-profile",         ["f"] = set_time_profile,         flags = {},             options = { "controller", "profile", "start-date", "end-date", "weekdays", "segments", "linked" } },
       { ["command"] = "delete-all-time-profiles", ["f"] = delete_all_time_profiles, flags = {},             options = { "controller", "profile" } },
       { ["command"] = "add-task",                 ["f"] = add_task,                 flags = {},             options = { "controller", "door", "task", "at", "start-date", "end-date", "weekdays", "more-cards" } },
       { ["command"] = "refresh-tasklist",         ["f"] = refresh_tasklist,         flags = {},             options = { "controller" } },
       { ["command"] = "clear-tasklist",           ["f"] = clear_tasklist,           flags = {},             options = { "controller" } },
       { ["command"] = "set-pc-control",           ["f"] = set_pc_control,           flags = { "disabled" }, options = { "controller" } },
       { ["command"] = "set-interlock",            ["f"] = set_interlock,            flags = {},             options = { "controller", "interlock" } },
       { ["command"] = "activate_keypads",         ["f"] = activate_keypads,         flags = {},             options = { "controller", "keypads" } },
       { ["command"] = "set-door-passcodes",       ["f"] = set_door_passcodes,       flags = {},             options = { "controller", "door", "passcodes" } },
       { ["command"] = "listen",                   ["f"] = listen,                   flags = {},             options = { "controller" } },
   },
}
-- stylua: ignore end

function commands.exec(cmd, args)
    for _,c in ipairs(commands.commands) do
        if c.command == cmd then
           exec(c,args)
           return
        end
    end

    error("unknown command:" .. cmd)
end

function exec(cmd, args)
    print()
    print(">> " .. cmd.command)
    print()

    local response = cmd.f(args)

    if cmd.command == "get-all-controllers" then
       for _, v in ipairs(response) do
           pprint(v)
       end
    elseif response then
       pprint(response)
    end
end


function pprint(v)
    local fields = v.fields()
    local w = 0

    for k,f in ipairs(fields) do
        if #f > w then
            w = #f
        end
    end

    local fmt = string.format("   %%-%ds  ",w)

    for k,f in ipairs(fields) do
        print(string.format(fmt,f) .. tostring(v[f]))
    end

    print()
end

function parse(args,key,defval)
    if args[key] ~= nil then
        return args[key]
    end

    return defval
end

return commands

File: bindings/erlang/src/commands.erl
-module(commands).

-export([commands/0, find/1, exec/3]).

-define(CONTROLLER, 405419896).
-define(DOOR, 3).
-define(MODE, 2).
-define(DELAY, 10).
-define(CARD, 10058400).
-define(CARD_INDEX, 1).
-define(EVENT_INDEX, 200).
-define(TIME_PROFILE_ID, 29).

-define(ADDRESS, "192.168.1.100").
-define(NETMASK, "255.255.255.0").
-define(GATEWAY, "192.168.1.1").
-define(LISTENER, "192.168.1.100:60001").

-define(LOG_TAG, "commands").

-include("records.hrl").

commands() ->
    [
        {"get-all-controllers", get_all_controllers},
        {"get-controller", get_controller},
        {"set-ip", set_ip},
        {"get-time", get_time},
        {"set-time", set_time},
        {"get-status", get_status},
        {"get-listener", get_listener},
        {"set-listener", set_listener},
        {"get-door-control", get_door_control},
        {"set-door-control", set_door_control},
        {"open-door", open_door},
        {"get-cards", get_cards},
        {"get-card", get_card},
        {"get-card-by-index", get_card_by_index},
        {"put-card", put_card},
        {"delete-card", delete_card},
        {"delete-all-cards", delete_all_cards},
        {"get-event", get_event},
        {"get-event-index", get_event_index},
        {"set-event-index", set_event_index},
        {"record-special-events", record_special_events},
        {"get-time-profile", get_time_profile},
        {"set-time-profile", set_time_profile},
        {"delete-all-time-profiles", delete_all_time_profiles},
        {"add-task", add_task},
        {"refresh-tasklist", refresh_tasklist},
        {"clear-tasklist", clear_tasklist},
        {"set-pc-control", set_pc_control},
        {"set-interlock", set_interlock},
        {"activate-keypads", activate_keypads},
        {"set-door-passcodes", set_door_passcodes},
        {"restore-default-parameters", restore_default_parameters},
        {"listen", listen}
    ].

find(Cmd) ->
    lists:keyfind(Cmd, 1, commands()).

exec({_, Cmd}, Options, Config) ->
    case execute(Cmd, Options, Config) of
        {ok, none} ->
            io:format("   ok~n");

        {ok, Any} ->
            pprint({ok, Any});

        Controllers when is_list(Controllers) ->
            lists:foreach(fun(X) -> pprint(X) end, Controllers);

        Other ->
            io:format(">> ~p~n", [Other])
    end.

execute(get_all_controllers, _Options, Config) ->
    uhppoted:get_all_controllers(Config);

execute(get_controller, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_controller(Config, Controller);

execute(set_ip, _Options, Config) ->
    Controller = ?CONTROLLER,
    {ok, Address} = inet:parse_ipv4_address(?ADDRESS),
    {ok, Netmask} = inet:parse_ipv4_address(?NETMASK),
    {ok, Gateway} = inet:parse_ipv4_address(?GATEWAY),

    uhppoted:set_ip(Config, Controller, Address, Netmask, Gateway);

execute(get_time, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_time(Config, Controller);

execute(set_time, _Options, Config) ->
    Controller = ?CONTROLLER,
    Now = erlang:localtime(),

    uhppoted:set_time(Config, Controller, Now);

execute(get_status, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_status(Config, Controller);

execute(get_listener, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_listener(Config, Controller);

execute(set_listener, _Options, Config) ->
    Controller = ?CONTROLLER,
    {Addr, Port} = parse_addr(?LISTENER),
    uhppoted:set_listener(Config, Controller, Addr, Port);

execute(get_door_control, _Options, Config) ->
    Controller = ?CONTROLLER,
    Door = ?DOOR,
    uhppoted:get_door_control(Config, Controller, Door);

execute(set_door_control, _Options, Config) ->
    Controller = ?CONTROLLER,
    Door = ?DOOR,
    Mode = ?MODE,
    Delay = ?DELAY,
    uhppoted:set_door_control(Config, Controller, Door, Mode, Delay);

execute(open_door, _Options, Config) ->
    Controller = ?CONTROLLER,
    Door = ?DOOR,
    uhppoted:open_door(Config, Controller, Door);

execute(get_cards, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_cards(Config, Controller);

execute(get_card, _Options, Config) ->
    Controller = ?CONTROLLER,
    Card = ?CARD,
    uhppoted:get_card(Config, Controller, Card);

execute(get_card_by_index, _Options, Config) ->
    Controller = ?CONTROLLER,
    Index = ?CARD_INDEX,
    uhppoted:get_card_by_index(Config, Controller, Index);

execute(put_card, _Options, Config) ->
    Controller = ?CONTROLLER,
    Card = ?CARD,
    Start = {2023, 1, 1},
    End = {2023, 12, 31},
    PIN = 7531,
    uhppoted:put_card(Config, Controller, Card, Start, End, 0, 1, 29, 0, PIN);

execute(delete_card, _Options, Config) ->
    Controller = ?CONTROLLER,
    Card = ?CARD,
    uhppoted:delete_card(Config, Controller, Card);

execute(delete_all_cards, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:delete_all_cards(Config, Controller);

execute(get_event, _Options, Config) ->
    Controller = ?CONTROLLER,
    Index = ?EVENT_INDEX,
    uhppoted:get_event(Config, Controller, Index);

execute(get_event_index, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:get_event_index(Config, Controller);

execute(set_event_index, _Options, Config) ->
    Controller = ?CONTROLLER,
    Index = ?EVENT_INDEX,
    uhppoted:set_event_index(Config, Controller, Index);

execute(record_special_events, _Options, Config) ->
    Controller = ?CONTROLLER,
    Enable = true,
    uhppoted:record_special_events(Config, Controller, Enable);

execute(get_time_profile, _Options, Config) ->
    Controller = ?CONTROLLER,
    Profile = ?TIME_PROFILE_ID,
    uhppoted:get_time_profile(Config, Controller, Profile);

execute(set_time_profile, _Options, Config) ->
    Controller = ?CONTROLLER,
    Profile = ?TIME_PROFILE_ID,
    Start = {2023, 1, 1},
    End = {2023, 12, 31},
    Monday = true,
    Tuesday = true,
    Wednesday = false,
    Thursday = true,
    Friday = false,
    Saturday = false,
    Sunday = true,
    Segment1Start = {8, 30},
    Segment1End = {11, 45},
    Segment2Start = {13, 15},
    Segment2End = {16, 30},
    Segment3Start = {19, 30},
    Segment3End = {20, 55},
    LinkedProfileID = 30,

    uhppoted:set_time_profile(
        Config,
        Controller,
        Profile,
        Start,
        End,
        Monday,
        Tuesday,
        Wednesday,
        Thursday,
        Friday,
        Saturday,
        Sunday,
        Segment1Start,
        Segment1End,
        Segment2Start,
        Segment2End,
        Segment3Start,
        Segment3End,
        LinkedProfileID
    );

execute(delete_all_time_profiles, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:delete_all_time_profiles(Config, Controller);

execute(add_task, _Options, Config) ->
    Controller = ?CONTROLLER,
    Door = ?DOOR,
    TaskType = 2,
    Start = {2023, 1, 1},
    End = {2023, 12, 31},
    Monday = true,
    Tuesday = false,
    Wednesday = true,
    Thursday = true,
    Friday = false,
    Saturday = false,
    Sunday = true,
    StartTime = {8, 30},
    MoreCards = 0,

    uhppoted:add_task(
        Config,
        Controller,
        Start,
        End,
        Monday,
        Tuesday,
        Wednesday,
        Thursday,
        Friday,
        Saturday,
        Sunday,
        StartTime,
        Door,
        TaskType,
        MoreCards
    );

execute(refresh_tasklist, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:refresh_tasklist(Config, Controller);

execute(clear_tasklist, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:clear_tasklist(Config, Controller);

execute(set_pc_control, _Options, Config) ->
    Controller = ?CONTROLLER,
    Enable = true,
    uhppoted:set_pc_control(Config, Controller, Enable);

execute(set_interlock, _Options, Config) ->
    Controller = ?CONTROLLER,
    Interlock = 3,
    uhppoted:set_interlock(Config, Controller, Interlock);

execute(activate_keypads, _Options, Config) ->
    Controller = ?CONTROLLER,
    Reader1 = true,
    Reader2 = true,
    Reader3 = false,
    Reader4 = true,
    uhppoted:activate_keypads(Config, Controller, Reader1, Reader2, Reader3, Reader4);

execute(set_door_passcodes, _Options, Config) ->
    Controller = ?CONTROLLER,
    Door = ?DOOR,
    Passcode1 = 12345,
    Passcode2 = 0,
    Passcode3 = 999999,
    Passcode4 = 54321,
    uhppoted:set_door_passcodes(Config, Controller, Door, Passcode1, Passcode2, Passcode3, Passcode4);

execute(restore_default_parameters, _Options, Config) ->
    Controller = ?CONTROLLER,
    uhppoted:restore_default_parameters(Config, Controller);

execute(listen, _Options, Config) ->
    case uhppoted:listen(Config, self()) of
        {ok, F} ->
            % in lieu of a CTRL-C handler (or more properly an OTP supervision tree)
            io:format("(type Q to quit)~n~n"),
            spawn(fun() ->
                io:fread("", "c"),
                F()
            end),
            listen();
        {error, Reason} ->
            {error, Reason}
    end;

execute(C, _, _) ->
    erlang:error({not_implemented, C}).

listen() ->
    receive
        {event, Event} ->
            pprint({event, Event}),
            listen();
        {error, Reason} ->
            log:errorf(?LOG_TAG, Reason),
            listen();
        closed ->
            log:infof(?LOG_TAG, closed)
    end.

parse_addr(S) ->
    [A, P] = string:tokens(S, ":"),
    {ok, Addr} = inet:parse_address(A),
    {Port, _} = string:to_integer(P),
    {Addr, Port}.

% Ref. https://erlang.org/pipermail/erlang-questions/2008-November/040029.html
pprint({ok, Response}) ->
    pretty_print(Response);

pprint({event, Event}) ->
    pretty_print(Event).

pretty_print(Record) ->
    [Name | Values] = tuple_to_list(Record),
    Fields = fields(Name),
    W = io_lib:format("       ~~-~Bs  ~~p~~n", [width(Fields)]),
    io:format("   ~s~n", [Name]),
    lists:foreach(fun({F, V}) -> io:format(W, [F, V]) end, zip(Fields, Values)),
    io:format("~n").

width(Fields) ->
    B = [atom_to_binary(X) || X <- Fields],
    N = [byte_size(X) || X <- B],
    lists:foldl(fun(X, W) -> max(X, W) end, 0, N).

zip(Fields, Values) ->
    zip(Fields, Values, []).

zip([], _, L) ->
    lists:reverse(L);

zip(_, [], L) ->
    lists:reverse(L);

zip([F | Fields], [V | Values], L) ->
    zip(Fields, Values, [{F, V} | L]).

{{range .model.responses}}
fields({{snakeCase .name}}) ->
    record_info(fields, {{snakeCase .name}});
{{end}}

{{with .model.event}}
fields({{snakeCase .name}}) ->
    record_info(fields, {{snakeCase .name}});
{{end}}

fields(_) ->
    [].



